lib/ansible/modules/network/fortimanager/fmgr_fwobj_service.py

```diff
@@ -0,0 +1,752 @@
+    meta fields = {}
+    color =
+    comment
+    """
+    response = (-100000, {"msg": "Nothing Happened."})
+    if paramgram["mode"] in ['set', 'add']:
+        url = '/pm/config/adom/{adom}/obj/firewall/service/group'.format(adom=paramgram["adom"])
+        datagram = {
+            "name": paramgram["group-name"],
+            "comment": paramgram["comment"],
+            "proxy": paramgram["explicit-proxy"],
+            "color": paramgram["color"]
+        }
+
+        members = paramgram["group-member"]
+        member = []
+        for obj in members.split(","):
+            member.append(obj.strip())
+        datagram["member"] = member
+
+    if paramgram["mode"] == "delete":
+        datagram = {
+            "name": paramgram["name"]
+        }
+        # SET DELETE URL
+        url = '/pm/config/adom/{adom}/obj/firewall/service/group' \
+              '/{name}'.format(adom=paramgram["adom"], name=paramgram["group-name"])
+
+    datagram = fmgr_del_none(datagram)
+
+    if paramgram["mode"] == "set":
+        response = fmg.set(url, datagram)
+        # IF MODE = ADD  -- USE THE 'ADD' API CALL MODE
+    if paramgram["mode"] == "add":
+        response = fmg.add(url, datagram)
+        # IF MODE = DELETE  -- USE THE DELETE URL AND API CALL MODE
+    if paramgram["mode"] == "delete":
+        response = fmg.delete(url, datagram)
+
+    return response
+
+
+def fmgr_fwobj_service_category(fmg, paramgram):
+    """
+    # NOTES
+    """
+    response = (-100000, {"msg": "Nothing Happened."})
+    if paramgram["mode"] in ['set', 'add']:
+        url = '/pm/config/adom/{adom}/obj/firewall/service/category'.format(adom=paramgram["adom"])
+        # GET RID OF ANY WHITESPACE
+        category = paramgram["category"]
+        category = category.strip()
+
+        datagram = {
+            "name": paramgram["category"],
+            "comment": "Created by Ansible"
+        }
+
+    # IF MODE = DELETE
+    if paramgram["mode"] == "delete":
+        datagram = {
+            "name": paramgram["name"]
+        }
+        # SET DELETE URL
+        url = '/pm/config/adom/{adom}/obj/firewall/service/category' \
+              '/{name}'.format(adom=paramgram["adom"], name=paramgram["category"])
+
+    datagram = fmgr_del_none(datagram)
+
+    if paramgram["mode"] == "set":
+        response = fmg.set(url, datagram)
+        # IF MODE = ADD  -- USE THE 'ADD' API CALL MODE
+    if paramgram["mode"] == "add":
+        response = fmg.add(url, datagram)
+        # IF MODE = DELETE  -- USE THE DELETE URL AND API CALL MODE
+    if paramgram["mode"] == "delete":
+        response = fmg.delete(url, datagram)
+
+    return response
+
+
+def fmgr_del_none(obj):
+    if isinstance(obj, dict):
+        return type(obj)((fmgr_del_none(k), fmgr_del_none(v))
+                         for k, v in obj.items() if k is not None and (v is not None and not fmgr_is_empty_dict(v)))
+    else:
+        return obj
+
+
+def fmgr_is_empty_dict(obj):
```

Comment:
I think you are trying to find a empty dict inside dict with max_depth=3 . better approach would be write a recursive function than to just write nested if/else loops

Reply:
Deepak, your input is appreciated.

However, in this case the max depth will never be more than 3. I could take the time to explain why, but know that it has to do with our process and how "deep" we allow modules to go in the depth of the API.